plugins/amdgpu/amdgpu_plugin_topology.c

```diff
@@ -0,0 +1,1104 @@
+		if (id_map->dest == dest_id)
+			return true;
+	}
+	return false;
+}
+
+uint32_t maps_get_dest_cpu(const struct device_maps *maps, const uint32_t src_id)
+{
+	struct id_map *id_map;
+	list_for_each_entry(id_map, &maps->cpu_maps, listm) {
+		if (id_map->src == src_id)
+			return id_map->dest;
+	}
+	return INVALID_CPU_ID;
+}
+
+bool maps_dest_gpu_mapped(const struct device_maps *maps, const uint32_t dest_id)
+{
+	struct id_map *id_map;
+	list_for_each_entry(id_map, &maps->gpu_maps, listm) {
+		if (id_map->dest == dest_id)
+			return true;
+	}
+	return false;
+}
+
+uint32_t maps_get_dest_gpu(const struct device_maps *maps, const uint32_t src_id)
+{
+	struct id_map *id_map;
+	list_for_each_entry(id_map, &maps->gpu_maps, listm) {
+		if (id_map->src == src_id)
+			return id_map->dest;
+	}
+	return 0;
+}
+
+struct id_map *maps_add_cpu_entry(struct device_maps *maps, const uint32_t src_id, const uint32_t dest_id)
+{
+	struct id_map *id_map = xmalloc(sizeof(*id_map));
+	if (!id_map) {
+		pr_err("Failed to allocate memory for id_map\n");
+		return NULL;
+	}
+	memset(id_map, 0, sizeof(*id_map));
+	id_map->src = src_id;
+	id_map->dest = dest_id;
+
+	list_add_tail(&id_map->listm, &maps->cpu_maps);
+
+	maps->tail_cpu = &id_map->listm;
+
+	pr_debug("Added CPU mapping [%02d -> %02d]\n", src_id, dest_id);
+	return id_map;
+}
+
+struct id_map *maps_add_gpu_entry(struct device_maps *maps, const uint32_t src_id, const uint32_t dest_id)
+{
+	struct id_map *id_map = xmalloc(sizeof(*id_map));
+	if (!id_map) {
+		pr_err("Failed to allocate memory for id_map\n");
+		return NULL;
+	}
+	memset(id_map, 0, sizeof(*id_map));
+	id_map->src = src_id;
+	id_map->dest = dest_id;
+
+	list_add_tail(&id_map->listm, &maps->gpu_maps);
+
+	maps->tail_gpu = &id_map->listm;
+
+	pr_debug("Added GPU mapping [0x%04X -> 0x%04X]\n", src_id, dest_id);
+	return id_map;
+}
+
+void maps_print(struct device_maps *maps)
+{
+	struct id_map *id_map;
+	pr_info("===Maps===============\n");
+	list_for_each_entry(id_map, &maps->gpu_maps, listm)
+		pr_info("GPU: 0x%04X -> 0x%04X\n", id_map->src, id_map->dest);
+
+	list_for_each_entry(id_map, &maps->cpu_maps, listm)
+		pr_info("CPU: 0x%02d -> 0x%02d\n", id_map->src, id_map->dest);
+	pr_info("======================\n");
+}
+
+void maps_init(struct device_maps *maps)
+{
+	INIT_LIST_HEAD(&maps->cpu_maps);
+	INIT_LIST_HEAD(&maps->gpu_maps);
+	maps->tail_cpu = 0;
+	maps->tail_gpu = 0;
+}
+
+void maps_free(struct device_maps *maps)
+{
+	while (!list_empty(&maps->cpu_maps)) {
+		struct id_map *map = list_first_entry(&maps->cpu_maps, struct id_map, listm);
+		list_del(&map->listm);
+		xfree(map);
+	}
+	while (!list_empty(&maps->gpu_maps)) {
+		struct id_map *map = list_first_entry(&maps->gpu_maps, struct id_map, listm);
+		list_del(&map->listm);
+		xfree(map);
+	}
+}
+
+void maps_remove(struct device_maps *maps, struct device_maps *remove)
+{
+	if (remove->tail_cpu)
+		list_cut_position(&remove->cpu_maps, &maps->cpu_maps, remove->tail_cpu);
```

Comment:
**maps_remove is the "reverse" of maps_append function.**

Example, if you have 3 lists:
list A = A1 - A2 - A3
list B = B1 - B2 - B3
list C = C1 - C2 - C3

Calling:
maps_append(A, B);
maps_append(A, C);

Will result in:
list A = A1 - A2 -A3 - B1 - B2 - B3 - C1 - C2 - C3

But maps_append will also store the location where the lists where glued together in tail_cpu and tail_gpu. So after maps_append(A, B) returns, the glue position will be stored in B.

So afterwards, if we call:

maps_remove(A, B);

Result will be:
list A= A1 - A2 - A3

i.e list A will be cut at the previous glue position stored in B and this result in elements B1 - B2 - B3 - C1 - C2 - C3 to be removed from A. 


**maps_remove implemetation:**
The list_cut_position function will split a list back into two lists at the previous glue position stored during maps_append.
Using the 3 lists in the previous example, after:

 list_cut_position(&remove->cpu_maps, &maps->cpu_maps, remove->tail_cpu)

remove->cpu_maps will contain elements B1 - B2 - B3 - C1 - C2 - C3
and 
maps->cpu_maps will contain elements A1 - A2 - A3

maps_remove then frees elements B1 - B2 - B3 - C1 - C2 - C3 by calling maps_free(remove)


For the maps->gpu_maps, we always add/remove one node at a time inside the map_device() function. But if one GPU has multiple iolinks to multiple CPUs, then for maps->cpu_maps, we would be appending multiple elements to maps->cpu_maps at a time, and if later on that combination fails, we have to remove multiple elements from maps->cpu_maps. This is why I am storing this glue position in maps_append().

Reply:
Looks to me like you store the "glue position" tail->cpu or tail->gpu in maps_add_cpu/gpu_entry, not maps_append.

It stores the last element added to the list. So the position stored in B would be B3, for example. But the glue position in your examples, as I would understand it, is A3.

I think you're using those lists like a stack during the recursion. You add things on the stack as you descend, and then pop them back off the stack as you come back up. Your maps_append function doesn't really "append" because it adds the new list entries at the start, not the end. Maybe that explains my confusion. With that in mind, the glue position makes a lot more sense. So the actual order of elements after maps_append(A, B) is B1 - B2 - B3 - A1 - A2 - A3, and B3 is where you want to split the list when you pop B with maps_remove(A, B).

I think renaming your maps_append/remove functions to maps_push/pop would go a long way to clearing that up. Also, add a comment explaining that these lists are used as stacks in the recursion.

Reply:
ACK. Yes, the intention was to make it behave like a stack.